src/wp-includes/bookmark.php

```diff
@@ -40,7 +40,13 @@ function get_bookmark( $bookmark, $output = OBJECT, $filter = 'raw' ) {
 			if ( ! $_bookmark ) {
 				$_bookmark = $wpdb->get_row( $wpdb->prepare( "SELECT * FROM $wpdb->links WHERE link_id = %d LIMIT 1", $bookmark ) );
 				if ( $_bookmark ) {
-					$_bookmark->link_category = array_unique( wp_get_object_terms( $_bookmark->link_id, 'link_category', array( 'fields' => 'ids' ) ) );
+					$link_category = wp_get_object_terms( $_bookmark->link_id, 'link_category', array( 'fields' => 'ids' ) );
+					if ( is_array( $link_category ) ) {
+						$_bookmark->link_category = array_unique( $link_category );
+					} else {
+						$_bookmark->link_category = null;
+						trigger_error( $link_category->get_error_message(), E_USER_WARNING );
```

Comment:
This code still constitutes a code assumption as while we know that `$link_category` is not an array, it has not been verified that it is an instance `WP_Error`, so we cannot be 100% sure that it can be approached as an object and that the `get_error_message()` method will be available.

This would only become problematic if `wp_get_object_terms()` would at some point change its return type (or if the currently documented return type is incomplete and we haven't realized that) or if the `WP_Error` class would change significantly, making the `get_error_message()` method (publicly) unavailable.

Reply:
>if the WP_Error class would change significantly, making the get_error_message() method (publicly) unavailable.
If `WP_Error::get_error_message()` becomes unavailable, this has breaking changes throughout Core, themes, and plugins. 

The likelihood of is very low; however, if this type of change happens, then the updates can be made here as well as throughout Core for the change.

>This code still constitutes a code assumption as while we know that $link_category is not an array, it has not been verified that it is an instance of WP_Error, so we cannot be 100% sure that it can be approached as an object and that the get_error_message() method will be available.

The more probable gotcha here is the filtered return value (at the end of the function). How so? If a callback does not return an array or an instance of `WP_Error`, then it fails into the `else` above and fails. 

But is that really a problem? The function has 2 designed data types: `array` and `WP_Error`. If a callback changes the data type, that failure will serve as an indication of something upstream is wrong.

>This would only become problematic if `wp_get_object_terms()` would at some point change its return type

This is true. If this function or any really changes it's return type(s), then it can be a breaking change downstream.

While that's true, could we go further to future proof the code to guard against future design changes? Let's unpack this. We'd have to assume another type other than `array` or `WP_Error` is returned. How do we know if that new future type is the desired alternative type or a failure? Hmm, we don't really.

Let's consider a 3rd decision branch:

```
if it's an array
elseif it's an instance of `WP_Errror`
else // future data type
```

Should we make that `else` a catchall? And if yes, then I'd advocate that we throw a warning to ensure upstream problems don't silently slip through the gates undetected.

What do you think @jrfnl?

Reply:
Hmm, thinking more about how to (a) protect the code and (b) capture an unexpected return type. 

The `else` is currently a catchall. We know the function should return an `array` or instance of `WP_Error`. However, we also know the final array is filtered before returning. There's the opportunity for failure.

With the current implementation, the severity of the failure is too high _and_ breaks from the _current_ behavior. Aha, this is an important point. Let's lean into this specific scenario.

What if.... walk with me ... what if, for this specific scenario, we trigger a less severe error such as a different warning message? 

This could look something like this:

```suggestion
						if ( $link_category instanceof WP_Error )
							$link_category = new WP_Error( 'invalid_return_type', __( 'Invalid return type from wp_get_objects_terms()' ) );	
						}
						trigger_error( $link_category->get_error_message(), E_USER_WARNING );
```

However, the above code only exists because any type could be returned from the filter, i.e. there are no strict restrictions. I'm thinking the above code would need to exist inside of `wp_get_objects_terms`, i.e. validate the returned type matches what's defined before returning it. Solve upstream instead of after the fact.

This train of thought loops back to the need for [Ticket #51525](https://core.trac.wordpress.org/ticket/51525).

The question for right now is this: 

- Do we guard the code here to prevent the possibility of a fatal error, i.e. which would be a breaking change?

Reply:
Been thinking about how to handle this, ie as @jrfnl points inspired a deeper probe. I wondered how it's used throughout Core. Why? The decisions we make here could be systematized for a consistent approach throughout this ticket.

Where to start? Audit code for the uses of `wp_get_object_terms`. The following audit was to discover: 

- if there are other expected types (than what the DocBlock identifies
- what the standard approach is for handling error vs array

Interesting results:

- Expected types: 
    - an instance of `WP_Error`
    - `array`
    - potentially an iterable (ie some implementations are not type casting when using return in a `foreach`). 
    - potentially mixed, as code does type casting
- Implementation varies from directly checking for an error to type casting to assuming an array is always returned.

The audit is below.

### Standard approach for handling returned types

There's a mix rather than a standard.

#### `is_wp_error` 

Checks if no error returned to continue; else skips to silently pass error through

- `wp-admin/nav-menus.php`:
     - [L79-81](https://github.com/WordPress/wordpress-develop/blob/ebe858e5a7b7900f5bc7bea955a69374c7cbb10f/src/wp-admin/nav-menus.php#L79)
     - [L156-159](https://github.com/WordPress/wordpress-develop/blob/ebe858e5a7b7900f5bc7bea955a69374c7cbb10f/src/wp-admin/nav-menus.php#L156)
- [`get_the_terms()`](https://github.com/WordPress/wordpress-develop/blob/97b2f07d2eba6cdd88bb960ba539d4c07f84f71f/src/wp-includes/category-template.php#L1279) => passes through a filter as `WP_Term[]|WP_Error`

#### Iterable

No type casting and used directly in `foreach`:

- [wp-includes/class-wp-xmlrpc-server.php](https://github.com/WordPress/wordpress-develop/blob/2ca15d1e5ce70493c5c0c096ca0c76503d6da07c/src/wp-includes/class-wp-xmlrpc-server.php#L916)
- [`get_terms_to_edit()`]( https://github.com/WordPress/wordpress-develop/blob/97b2f07d2eba6cdd88bb960ba539d4c07f84f71f/src/wp-admin/includes/taxonomy.php#L266) passed into `wp_list_pluck`
- [`get_attachment_fields_to_edit()`](https://github.com/WordPress/wordpress-develop/blob/5bc27786d1568f68bd2c7d4a2f8230c73fc18a28/src/wp-admin/includes/media.php#L1439)
- [`get_compat_media_markup()`](https://github.com/WordPress/wordpress-develop/blob/5bc27786d1568f68bd2c7d4a2f8230c73fc18a28/src/wp-admin/includes/media.php#L1898)

#### Array

Assumes `array` is returned:
- [`wp_delete_object_term_relationships()`](https://github.com/WordPress/wordpress-develop/blob/aac637dcdc690050f51cc7cb20f1230bafe4a98d/src/wp-includes/taxonomy.php#L1800) passed into `array_map()`.
- [`wp_get_link_cats()`](https://github.com/WordPress/wordpress-develop/blob/8d5bc0835b25255b121aa218ae2b174359aa4784/src/wp-admin/includes/bookmark.php#L128) passed into `array_map`
- [`WP_REST_Terms_Controller::get_items()`](https://github.com/WordPress/wordpress-develop/blob/0228dd6a5d17aa42735fdff9b106afccb960311e/src/wp-includes/rest-api/endpoints/class-wp-rest-terms-controller.php#L254)
- [`wp_terms_checklist()`](https://github.com/WordPress/wordpress-develop/blob/5bc27786d1568f68bd2c7d4a2f8230c73fc18a28/src/wp-admin/includes/template.php#L125) used in `in_array()`
- [`wp_popular_terms_checklist()`](https://github.com/WordPress/wordpress-develop/blob/5bc27786d1568f68bd2c7d4a2f8230c73fc18a28/src/wp-admin/includes/template.php#L213) used in `in_array()`
- [`get_adjacent_post()`](https://github.com/WordPress/wordpress-develop/blob/8b2aa594e39f170452f4aab138273cf3af0ff26e/src/wp-includes/link-template.php#L1771) passes into `array_diff`
- [`get_boundary_post()`](https://github.com/WordPress/wordpress-develop/blob/8b2aa594e39f170452f4aab138273cf3af0ff26e/src/wp-includes/link-template.php#L2063) passes into `array_diff`

Checks `is_array()`:
- [`_wp_menu_item_classes_by_context()`](https://github.com/WordPress/wordpress-develop/blob/8b89be2676380d67e8e6df1568d736b9faab04d9/src/wp-includes/nav-menu-template.php#L332)

#### Type cast to Array

- [`export_wp()`](https://github.com/WordPress/wordpress-develop/blob/97b2f07d2eba6cdd88bb960ba539d4c07f84f71f/src/wp-admin/includes/export.php#L418)
- `bulk_edit_posts()`: [L601](https://github.com/WordPress/wordpress-develop/blob/8e25b92b63514a5ce7518de09811c58866e286fa/src/wp-admin/includes/post.php#L601) and [L603](https://github.com/WordPress/wordpress-develop/blob/8e25b92b63514a5ce7518de09811c58866e286fa/src/wp-admin/includes/post.php#L603)

#### Returns what is returned

- [`wp_get_post_categories()`](https://github.com/WordPress/wordpress-develop/blob/637a3e8b5dfad3e30fdbe38a2547cdbe8eb5499e/src/wp-includes/post.php#L3495)

Reply:
Code review notes from discussion with @jrfnl:

- Remove the `trigger_error` as this warning does not add value in this instance
- `null` is the expected value when not an `array`. This is correct.
- Create a test return to a different type other than `array` or `WP_Error` by hooking into the `wp_get_object_terms` filter.
- This is a perfect example of why we need strict return filtered value https://core.trac.wordpress.org/ticket/51525. Make sure this ticket is linked to Ticket 51525.
- Make sure the comments and reasoning appear in ticket itself.